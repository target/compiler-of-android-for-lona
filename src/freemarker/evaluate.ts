import util from 'util'
import escapeRegExp from 'lodash.escaperegexp'
import ProgramNode from 'freemarker-parser/src/nodes/ProgramNode'
import TextNode from 'freemarker-parser/src/nodes/TextNode'
import ConditionNode from 'freemarker-parser/src/nodes/ConditionNode'
import InterpolationNode from 'freemarker-parser/src/nodes/InterpolationNode'
import ListNode from 'freemarker-parser/src/nodes/ListNode'
import AbstractBodyNode from 'freemarker-parser/src/nodes/abstract/AbstractBodyNode'
import ParamNames from 'freemarker-parser/src/enum/ParamNames'
import {
  Compound,
  Expression,
  Identifier,
  Literal,
  CallExpression,
  UnaryExpression,
  LogicalExpression,
  BuiltInExpression,
  BinaryExpression,
} from 'freemarker-parser/src/interface/Params'

export type Context = {
  data: any
}

export function evaluateCondition(
  node: ConditionNode,
  context: Context
): string {
  const { params } = node

  return params && evaluateExpression(params, context)
    ? evaluateBody(node.consequent, context)
    : node.alternate
    ? evaluateBody(node.alternate, context)
    : ''
}

export function evaluateExpression(
  unclassified: Expression,
  context: Context
): any {
  const { type } = unclassified

  switch (type) {
    case 'Identifier': {
      const expression = unclassified as Identifier

      if (!(expression.name in context.data)) {
        console.log('missing field', expression.name)
      }

      return context.data[expression.name]
    }
    case 'Literal': {
      const expression = unclassified as Literal

      return expression.value
    }
    case 'UnaryExpression': {
      const expression = unclassified as UnaryExpression
      switch (expression.operator) {
        case '??':
          return !!evaluateExpression(expression.argument, context)
        case '!':
          return !evaluateExpression(expression.argument, context)
      }
      console.log('Unhandled unary expression', expression)
      return null
    }
    case 'CallExpression': {
      const expression = unclassified as CallExpression
      const callee = evaluateExpression(expression.callee, context)
      const args = expression.arguments.map(arg =>
        evaluateExpression(arg, context)
      )
      if (typeof callee === 'function') {
        return callee.apply(context, args)
      }
      console.error(
        `Attempted to call non-function: ${callee}\n\nAST:\n${JSON.stringify(
          expression.callee
        )}`
      )
      return null
    }
    case 'BuiltInExpression': {
      const expression = unclassified as BuiltInExpression
      const { left, right, operator } = expression
      // const name = (left as Identifier).name as BuiltInMethodName

      const method: CallExpression =
        right.type === ParamNames.Identifier
          ? {
              type: ParamNames.CallExpression,
              callee: right,
              arguments: [],
            }
          : (right as CallExpression)

      const name = (method.callee as Identifier).name

      if (!name || operator !== '?') {
        console.error('Unhandled built-in expression', unclassified)
        return null
      }

      // const evaluatedObject = evaluateExpression(left, context)
      // const evaluatedArguments = method.arguments.map(arg =>
      //   evaluateExpression(arg, context)
      // )

      switch (name) {
        case 'replace':
          const value = evaluateExpression(left, context)
          const [original, replacement, flags] = method.arguments.map(arg =>
            evaluateExpression(arg, context)
          )
          const result = value.replace(
            new RegExp(escapeRegExp(original), flags),
            replacement
          )
          return result
        case 'has_content': {
          const value = evaluateExpression(left, context)
          const typeName = {}.toString.call(value).slice(8, -1)
          switch (typeName) {
            case 'String':
              return value !== ''
            case 'Boolean':
              return value !== false
            case 'Number':
              return value !== 0
            case 'Array':
              return value.length !== 0
            default:
              return !!typeName
          }
        }
        case 'string': {
          return evaluateExpression(left, context).toString()
        }
        case 'matches': {
          const value = evaluateExpression(left, context)
          const raw = (method.arguments[0] as Literal).raw
          const regexp = new RegExp(JSON.parse(raw))
          // console.log(
          //   'MATCHES',
          //   util.inspect(unclassified, false, null, true),
          //   regexp,
          //   raw,
          //   evaluatedObject
          // )
          return regexp.test(value)
        }
        default:
          console.error('Unhandled built-in expression', name)
          return null
      }
    }
    case 'BinaryExpression': {
      const expression = unclassified as BinaryExpression
      const { left, right, operator } = expression
      switch (operator) {
        case '!=':
          return !!(
            evaluateExpression(left, context) !=
            evaluateExpression(right, context)
          )
        case '==':
          return !!(
            evaluateExpression(left, context) ==
            evaluateExpression(right, context)
          )
        case 'gt':
        case '>':
          return (
            evaluateExpression(left, context) >
            evaluateExpression(right, context)
          )
        case 'gte':
        case '>=':
          return (
            evaluateExpression(left, context) >=
            evaluateExpression(right, context)
          )
        case 'lt':
        case '<':
          return (
            evaluateExpression(left, context) <
            evaluateExpression(right, context)
          )
        case 'lte':
        case '<=':
          return (
            evaluateExpression(left, context) <=
            evaluateExpression(right, context)
          )
        default:
          console.error('Unhandled binary expression', unclassified)
      }
    }
    case 'LogicalExpression': {
      const expression = unclassified as LogicalExpression
      const { left, right, operator } = expression
      switch (operator) {
        case '||':
          return !!(
            evaluateExpression(left, context) ||
            evaluateExpression(right, context)
          )
        case '&&':
          return !!(
            evaluateExpression(left, context) &&
            evaluateExpression(right, context)
          )
        default:
          console.error('Unhandled logical expression', unclassified)
      }
    }
    default:
      console.error('Unhandled expression', unclassified)
  }
}

// InterpolationNode {
//   type: 'Interpolation',
//   params: {
//     type: 'CallExpression',
//     arguments: [ { type: 'Identifier', name: 'topOut' } ],
//     callee: { type: 'Identifier', name: 'escapeXmlAttribute' }
//   }
// },
export function evaluateInterpolation(
  node: InterpolationNode,
  context: Context
): string {
  return node.params ? evaluateExpression(node.params, context) : ''
}

// {
//   type: 'List',
//   params: {
//     type: 'Compound',
//     body: [
//       { type: 'Identifier', name: 'dependencyList' },
//       { type: 'Identifier', name: 'as' },
//       { type: 'Identifier', name: 'dependency' }
//     ]
//   },
//   body: [],
// }
export function evaluateList(node: ListNode, context: Context): string {
  if (node.params?.type !== ParamNames.Compound) {
    console.error('Bad list node', node)
    return ''
  }

  const compound = node.params as Compound
  const identifiers = compound.body as Identifier[]

  const [
    { name: variable },
    { name: operator },
    { name: identifier },
  ] = identifiers

  if (operator !== 'as') {
    console.error('Bad list node', node)
    return ''
  }

  // Set up child scope
  const scopeValue = context.data[identifier]
  context.data[identifier] = variable

  const result = evaluateBody(node.body, context)

  // Restore parent scope
  context.data[identifier] = scopeValue

  return result
}

export function evaluateBody(
  nodes: AbstractBodyNode[],
  context: Context
): string {
  return nodes
    .map(unclassified => {
      switch (unclassified.type) {
        case 'Text': {
          const node = unclassified as TextNode
          return node.text
        }
        case 'Condition': {
          const node = unclassified as ConditionNode
          return evaluateCondition(node, context)
        }
        case 'Interpolation': {
          const node = unclassified as InterpolationNode
          return evaluateInterpolation(node, context)
        }
        case 'List': {
          const node = unclassified as ListNode
          return evaluateList(node, context)
        }
        default:
          console.error(
            `Node not handled: ${
              unclassified
              // util.inspect(unclassified, false, null, true)
            }`
          )
          return ''
      }
    })
    .join('')
}

export function evaluate(program: ProgramNode, context: Context): string {
  return evaluateBody(program.body || [], context)
}

export { ProgramNode }
