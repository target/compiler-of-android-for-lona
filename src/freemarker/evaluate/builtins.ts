import escapeRegExp from 'lodash.escaperegexp'

function typeName(value: any): string {
  return {}.toString.call(value).slice(8, -1)
}

/**
 * Reference:
 * https://freemarker.apache.org/docs/ref_builtins_expert.html#ref_builtin_has_content
 */
export function hasContent(instance: any, args: any[]) {
  switch (typeName(instance)) {
    case 'String':
      return instance !== ''
    case 'Boolean':
      return instance !== false
    case 'Number':
      return instance !== 0
    case 'Array':
      return instance.length !== 0
    case 'Object':
      return Object.keys(instance).length !== 0
    case 'Null':
      return false
    case 'Undefined':
      return false
    default:
      return !!typeName
  }
}

/**
 * Reference:
 * https://freemarker.apache.org/docs/ref_builtins_string.html#ref_builtin_replace
 *
 * For flags, see:
 * https://freemarker.apache.org/docs/ref_builtins_string.html#ref_builtin_string_flags
 */
export function replace(instance: any, args: any[]) {
  const [original, replacement, flags] = args

  const firstOnly = flags.includes('f')
  const regExpFlags = firstOnly ? flags.replace('f', '') : flags + 'g'

  const result = instance.replace(
    new RegExp(escapeRegExp(original), regExpFlags),
    replacement
  )

  return result
}

/**
 * Reference (multiple definitions):
 * https://freemarker.apache.org/docs/ref_builtins_alphaidx.html
 */
export function string(instance: any, args: any[]) {
  return instance.toString()
}

/**
 * Reference:
 * https://freemarker.apache.org/docs/ref_builtins_string.html#ref_builtin_matches
 */
export function matches(instance: any, args: any[]) {
  const escaped = '"' + args[0].replace(/"/g, '\\"') + '"'
  const regexp = new RegExp(JSON.parse(escaped))
  return regexp.test(instance)
}

export function evaluateBuiltInMethod(
  methodName: string,
  instance: any,
  args: any[]
): any {
  switch (methodName) {
    case 'replace':
      return replace(instance, args)
    case 'has_content':
      return hasContent(instance, args)
    case 'string':
      return string(instance, args)
    case 'matches':
      return matches(instance, args)
    default:
      console.error('Unhandled built-in method:', methodName)
      return undefined
  }
}
