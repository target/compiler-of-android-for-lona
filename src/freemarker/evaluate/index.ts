import ProgramNode from 'freemarker-parser/dist/nodes/ProgramNode'
import { Context } from '../context'
import { evaluateNodes } from './node'

export function evaluate(program: ProgramNode, context: Context): string {
  return evaluateNodes(program.body || [], context)
}

export { ProgramNode }
