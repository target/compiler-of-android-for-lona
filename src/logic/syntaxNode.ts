import {
  TraversalConfig,
  emptyConfig,
  reduce,
} from '@lona/compiler/lib/helpers/logic-traversal'
import { AST } from '@lona/compiler/lib/helpers/logic-ast'

// TODO: Reduce isn't returning a value correctly
export function findNode(
  rootNode: AST.SyntaxNode,
  predicate: (node: AST.SyntaxNode) => boolean
): AST.SyntaxNode | undefined {
  let found: AST.SyntaxNode | undefined

  reduce(
    rootNode,
    (_result, node, config) => {
      if (!found && predicate(node)) {
        config.stopTraversal = true
        found = node
      }
      return undefined
    },
    undefined
  )

  return found
}

export function forEach(
  rootNode: AST.SyntaxNode,
  config: TraversalConfig,
  f: (node: AST.SyntaxNode, config: TraversalConfig) => void
) {
  reduce(
    rootNode,
    (_result, node, config) => {
      f(node, config)
      return undefined
    },
    undefined,
    config
  )
}

export const Traversal = {
  get preorder(): TraversalConfig {
    return {
      ...emptyConfig(),
      order: 'PreOrder' as TraversalConfig['order'],
    }
  },
}
