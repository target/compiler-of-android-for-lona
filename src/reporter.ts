import { Reporter } from '@lona/compiler/lib/helpers/reporter'

export const silentReporter: Reporter = {
  log: () => {},
  info: () => {},
  warn: () => {},
  error: () => {},
}
