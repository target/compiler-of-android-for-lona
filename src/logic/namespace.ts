import {
  TraversalConfig,
  emptyConfig,
  reduce,
} from '@lona/compiler/lib/helpers/logic-traversal'
import { LogicAST as AST } from '@lona/serialization'
import { NodePath } from './nodePath'
import { createDeclarationNode } from './nodes/declarations'
import { forEach, Traversal, visit } from './syntaxNode'

export type UUID = string

export type Namespace = {
  values: { [key: string]: UUID }
  types: { [key: string]: UUID }
}

export const builtInTypeConstructorNames: Set<String> = new Set([
  'Boolean',
  'Number',
  'String',
  'Array',
  'Color',
])

/**
 * Merge namespaces, throwing an error in the case of collisions
 */
export function merge(namespaces: Namespace[]): Namespace {
  return namespaces.reduce((result, namespace) => {
    Object.entries(namespace.values).forEach(([key, value]) => {
      if (key in result.values) {
        throw new Error(`Namespace error: value ${key} declared more than once`)
      }

      result.values[key] = value
    })

    Object.entries(namespace.types).forEach(([key, type]) => {
      if (key in result.types) {
        throw new Error(`Namespace error: type ${key} declared more than once`)
      }

      result.types[key] = type
    })

    return result
  }, createNamespace())
}

export function copy(namespace: Namespace): Namespace {
  return {
    values: { ...namespace.values },
    types: { ...namespace.types },
  }
}

export class NamespaceVisitor {
  namespace: Namespace
  currentPath = new NodePath()

  constructor(namespace: Namespace) {
    this.namespace = namespace
  }

  pushPathComponent(name: string) {
    this.currentPath.pushComponent(name)
  }

  popPathComponent() {
    this.currentPath.popComponent()
  }

  declareValue(name: string, value: UUID) {
    const path = this.currentPath.pathString(name)

    if (this.namespace.values[path]) {
      throw new Error(`Value already declared: ${path}`)
    }

    this.namespace.values[path] = value
  }

  declareType(name: string, type: UUID) {
    const path = this.currentPath.pathString(name)

    if (this.namespace.types[path]) {
      throw new Error(`Type already declared: ${path}`)
    }

    this.namespace.types[path] = type
  }

  traverse(rootNode: AST.SyntaxNode) {
    visit(rootNode, Traversal.preorder, {
      enter: node => createDeclarationNode(node)?.namespaceEnter(this),
      leave: node => createDeclarationNode(node)?.namespaceLeave(this),
    })
  }
}

/**
 * Build the global namespace by visiting each node.
 */
export function createNamespace(topLevelNode?: AST.SyntaxNode): Namespace {
  let namespace: Namespace = { types: {}, values: {} }

  if (topLevelNode) {
    let visitor = new NamespaceVisitor(namespace)

    visitor.traverse(topLevelNode)
  }

  return namespace
}
