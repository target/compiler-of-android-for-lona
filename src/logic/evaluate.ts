import { declarationPathTo } from '@lona/compiler/lib/helpers/logic-ast'
import { Reporter } from '@lona/compiler/lib/helpers/reporter'
import { assertNever } from '@lona/compiler/lib/utils/assert-never'
import { nonNullable } from '@lona/compiler/lib/utils/non-nullable'
import { ShallowMap } from '@lona/compiler/lib/utils/shallow-map'
import { LogicAST as AST } from '@lona/serialization'
import { Value, StandardLibrary } from './runtime/value'
import { Scope } from './scope'
import * as StaticType from './staticType'
import { TypeChecker } from './typeChecker'
import { substitute, Substitution } from './typeUnifier'
import { createNode, createEvaluationVisitor } from './nodes/createNode'

const STANDARD_LIBRARY = 'standard library'

export function evaluateIsTrue(
  context: EvaluationContext,
  expression: AST.Expression
) {
  const condition = context.evaluate(expression.data.id)
  return (
    (condition &&
      condition.type.type === 'constant' &&
      condition.type.name === 'Boolean' &&
      condition.memory.type === 'bool' &&
      condition.memory.value) ||
    false
  )
}

type Thunk = {
  label: string
  dependencies: string[]
  f: (args: Value[]) => Value
}

export class EvaluationVisitor {
  evaluation: EvaluationContext
  rootNode: AST.SyntaxNode
  scope: Scope
  reporter: Reporter
  typeChecker: TypeChecker
  substitution: Substitution

  constructor(
    rootNode: AST.SyntaxNode,
    scope: Scope,
    typeChecker: TypeChecker,
    substitution: Substitution,
    reporter: Reporter
  ) {
    this.evaluation = new EvaluationContext(reporter)
    this.rootNode = rootNode
    this.scope = scope
    this.reporter = reporter
    this.typeChecker = typeChecker
    this.substitution = substitution
  }

  add(uuid: string, thunk: Thunk) {
    this.evaluation.add(uuid, thunk)
  }

  addValue(uuid: string, value: Value) {
    this.evaluation.addValue(uuid, value)
  }
}

/**
 * The evaluation context of the Lona Workspace.
 */
export class EvaluationContext {
  values: { [uuid: string]: Value } = {}
  thunks: { [uuid: string]: Thunk } = {}
  reporter?: Reporter

  constructor(reporter?: Reporter) {
    this.reporter = reporter
  }

  add(uuid: string, thunk: Thunk) {
    this.thunks[uuid] = thunk
  }

  addValue(uuid: string, value: Value) {
    this.values[uuid] = value
  }

  /**
   * Evaluate the id to a value, resolving any dependency along the way
   */
  evaluate(uuid: string): Value | undefined {
    const value = this.values[uuid]

    if (value) return value

    const thunk = this.thunks[uuid]

    if (!thunk) {
      this.reporter?.error(`no thunk for ${uuid}`)
      return undefined
    }

    const resolvedDependencies = thunk.dependencies.map(x => this.evaluate(x))

    if (resolvedDependencies.some(x => !x)) {
      this.reporter?.error(
        `Failed to evaluate thunk ${uuid} (${thunk.label}) - missing dep ${
          thunk.dependencies[resolvedDependencies.findIndex(x => !x)]
        }`
      )
      return undefined
    }

    const result = thunk.f(resolvedDependencies as Value[])
    this.values[uuid] = result

    return result
  }

  copy() {
    const newContext = new EvaluationContext(this.reporter)
    newContext.thunks = { ...this.thunks }
    newContext.values = { ...this.values }
    return newContext
  }
}

export const evaluate = (
  node: AST.SyntaxNode,
  visitor: EvaluationVisitor
): EvaluationContext | undefined => {
  const {
    rootNode,
    scope,
    typeChecker,
    substitution,
    reporter,
    evaluation: initialContext,
  } = visitor

  // TODO: Handle stopping
  const context = AST.subNodes(node).reduce<EvaluationContext | undefined>(
    (prev, subNode) => {
      if (!prev) {
        return undefined
      }
      return evaluate(subNode, visitor)
    },
    initialContext
  )

  if (!context) {
    return undefined
  }

  switch (node.type) {
    case 'identifierExpression':
    case 'none':
    case 'boolean':
    case 'number':
    case 'string':
    case 'color':
    case 'array':
    case 'literalExpression':
    case 'memberExpression':
    case 'record':
    case 'variable':
    case 'enumeration':
    case 'function':
    case 'functionCallExpression':
      const visitorNode = createEvaluationVisitor(node)

      if (visitorNode) {
        visitorNode.evaluationEnter(visitor)
      }

      break
    case 'assignmentExpression': {
      visitor.add(node.data.left.data.id, {
        label:
          'Assignment for ' +
          declarationPathTo(rootNode, node.data.left.data.id).join('.'),
        dependencies: [node.data.right.data.id],
        f: values => values[0],
      })
      break
    }
    case 'functionType':
    case 'typeIdentifier':
    case 'program':
    case 'parameter':
    case 'value':
    case 'topLevelParameters':
    case 'topLevelDeclarations':
    case 'enumerationCase': // handled in 'enumeration'
    case 'argument': // handled in 'functionCallExpression'
    case 'namespace':
    case 'importDeclaration':
    case 'placeholder':
    case 'return': // handled in 'function'
    case 'loop': // handled in 'function'
    case 'branch': // handled in 'function'
    case 'expression':
    case 'declaration': {
      break
    }
    default: {
      assertNever(node)
    }
  }

  return context
}
