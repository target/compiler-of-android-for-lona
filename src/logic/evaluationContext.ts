import * as fs from 'fs'
import * as path from 'path'
import { LogicAST as AST, decodeLogic } from '@lona/serialization'
import * as LogicScope from './scope'
import * as LogicUnify from './typeUnifier'
import * as LogicEvaluate from './evaluate'
import { Reporter } from '@lona/compiler/lib/helpers/reporter'
import { nonNullable } from '@lona/compiler/lib/utils/non-nullable'
import { makeProgram, joinPrograms } from '@lona/compiler/lib/helpers/logic-ast'
import { createNamespace } from './namespace'
import { makeUnificationContext } from './typeChecker'

export const STANDARD_LIBRARY = 'standard library'

export const generate = (reporter: Reporter, programs: AST.SyntaxNode[]) => {
  const standardLibsPath = path.join(__dirname, 'library')
  const standardLibs = fs.readdirSync(standardLibsPath)

  const libraryFiles: AST.Program[] = standardLibs.map(
    x =>
      makeProgram(
        decodeLogic(fs.readFileSync(path.join(standardLibsPath, x), 'utf8'))
      ) as AST.Program
  )

  const standardLibsProgram = joinPrograms(libraryFiles)

  const logicPrograms = programs.map(makeProgram).filter(nonNullable)

  const programNode = joinPrograms([standardLibsProgram, ...logicPrograms])

  const namespace = createNamespace(programNode)
  const scopeContext = LogicScope.createScopeContext(programNode, namespace)

  const unificationContext = makeUnificationContext(
    programNode,
    scopeContext,
    reporter
  )
  const substitution = LogicUnify.unify(
    unificationContext.constraints,
    reporter
  )

  const evaluationContext = LogicEvaluate.evaluate(
    programNode,
    programNode,
    scopeContext,
    unificationContext,
    substitution,
    reporter
  )

  return evaluationContext
}
