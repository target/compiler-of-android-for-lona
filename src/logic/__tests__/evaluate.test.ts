import { AST } from '@lona/compiler/lib/helpers/logic-ast'
import * as Serialization from '@lona/serialization'
import { silentReporter } from '../../reporter'
import { EvaluationContext } from '../evaluate'
import { generate } from '../evaluationContext'
import { UUID } from '../namespace'
import { findNode } from '../syntaxNode'

function getInitializerId(
  rootNode: AST.SyntaxNode,
  variableName: string
): UUID {
  const variable = findNode(rootNode, node => {
    return node.type === 'variable' && node.data.name.name === variableName
  }) as AST.VariableDeclaration | undefined

  if (!variable) {
    throw new Error(`Variable ${variableName} not found`)
  }

  const initializer = variable.data.initializer

  if (!initializer) {
    throw new Error(`Initializer for ${variableName} not found`)
  }

  return initializer.data.id
}

function standardEvaluate(rootNode: AST.SyntaxNode): EvaluationContext {
  const evaluation = generate(console, [rootNode])

  if (!evaluation) {
    throw new Error('Failed to evaluate')
  }

  return evaluation
}

describe('Logic / Evaluate', () => {
  it('evaluates number literals', () => {
    const file = `let x: Number = 4`
    const rootNode = Serialization.decodeLogic(file)
    const initializerId = getInitializerId(rootNode, 'x')
    const evaluation = standardEvaluate(rootNode)

    expect(evaluation.evaluate(initializerId)).toEqual({
      type: { type: 'constant', name: 'Number', parameters: [] },
      memory: { type: 'number', value: 4 },
    })
  })

  it('evaluates number literals', () => {
    const file = `
enum Foo {
  case bar()
}

let x: Foo = Foo.bar()
    `
    const rootNode = Serialization.decodeLogic(file)
    const initializerId = getInitializerId(rootNode, 'x')
    const evaluation = standardEvaluate(rootNode)

    expect(evaluation.evaluate(initializerId)).toEqual({
      type: { type: 'constant', name: 'Foo', parameters: [] },
      memory: { type: 'enum', value: 'bar', data: [] },
    })
  })

  it('evaluates DimensionSize', () => {
    const file = `
let x: DimensionSize = DimensionSize.fixed(100)
`
    const rootNode = Serialization.decodeLogic(file)
    const initializerId = getInitializerId(rootNode, 'x')
    const evaluation = standardEvaluate(rootNode)

    expect(evaluation.evaluate(initializerId)).toMatchSnapshot()
  })

  it('evaluates custom function', () => {
    const file = `
func test() -> DimensionSize {
  return DimensionSize.fixed(100)
}

let x: DimensionSize = test()
`
    const rootNode = Serialization.decodeLogic(file)
    const initializerId = getInitializerId(rootNode, 'x')
    const evaluation = standardEvaluate(rootNode)

    const result = evaluation.evaluate(initializerId)

    expect(result).toMatchSnapshot()
  })

  //   it('evaluates View function', () => {
  //     const file = `
  // let x: Element = View()
  // `
  //     const rootNode = Serialization.decodeLogic(file)
  //     const initializerId = getInitializerId(rootNode, 'x')
  //     const evaluation = standardEvaluate(rootNode)

  //     const result = evaluation.evaluate(initializerId)

  //     expect(result).toEqual({})
  //   })
})
