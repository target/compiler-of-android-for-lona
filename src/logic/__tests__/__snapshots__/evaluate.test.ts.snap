// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logic / Evaluate evaluates DimensionSize 1`] = `
Object {
  "memory": Object {
    "data": Array [
      Object {
        "memory": Object {
          "type": "number",
          "value": 100,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
    ],
    "type": "enum",
    "value": "fixed",
  },
  "type": Object {
    "name": "DimensionSize",
    "parameters": Array [],
    "type": "constant",
  },
}
`;

exports[`Logic / Evaluate evaluates DimensionSize 2`] = `
Object {
  "memory": Object {
    "data": Array [
      Object {
        "memory": Object {
          "type": "number",
          "value": 100,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
    ],
    "type": "enum",
    "value": "fixed",
  },
  "type": Object {
    "name": "DimensionSize",
    "parameters": Array [],
    "type": "constant",
  },
}
`;

exports[`Logic / Evaluate evaluates View function 1`] = `undefined`;

exports[`Logic / Evaluate evaluates custom function 1`] = `
Object {
  "memory": Object {
    "data": Array [
      Object {
        "memory": Object {
          "type": "number",
          "value": 100,
        },
        "type": Object {
          "name": "Number",
          "parameters": Array [],
          "type": "constant",
        },
      },
    ],
    "type": "enum",
    "value": "fixed",
  },
  "type": Object {
    "name": "DimensionSize",
    "parameters": Array [],
    "type": "constant",
  },
}
`;
